make check-parallel:  Parallel 'make check'

gc00 opened the pull request:
I'm still testing on this one.  But `make check-parallel` should work now.  Similarly, `autotest.py -j` and `atuotest.py --parallel` should now work.

gc00 committed:
make check-parallel:  Parallel 'make check'

karya0 commented on test/autotest.py:
```diff
@@ -292,7 +360,8 @@ def runCmd(cmd):
 os.mkdir(ckptDir);
 os.environ['DMTCP_COORD_HOST'] = "localhost"
-os.environ['DMTCP_COORD_PORT'] = str(randint(2000,10000))
+if not os.getenv('DMTCP_COORD_PORT'):
```
We need to revert this change. I often have a coordinator already running at `DMTCP_COORD_PORT` (which is defined in the environment). With this change, the tests will try to connect to the existing coordinator.

Instead, we should make it something like:

    if not os.getenv('DMTCP_NUM_RUN'):
      os.environ['DMTCP_COORD_PORT'] = str(randint(2000,10000))

gc00 replied:
Will do.

gc00 committed:
enable parallel --slow

gc00 commented:
I added `DMTCP_PARALLEL_AUTOTEST` environment variable to distinguish the two levels (top-level autotest and parallel autotest on indidvidual test).  This also fixes the inlilne request by @karya0.

It still creates lots of directores `dmtcp-autotest-*` that are not cleaned up.  I still have to fix that.

rohgarg commented on Makefile.in:
```diff
@@ -143,6 +143,9 @@ check: tests
 	   ' python to path or build Python 2.x for x >= 3.'; \
 	  fi
 
+check-parallel: tests
```
Can we use the `make tidy` target to clean up the `autotest-*` files?

gc00 committed:
Kill stray dmtcp_coordinator and test processes

gc00 commented:
@rohgarg and @karya0, We'll have to wait until @karya0 approves the changes to unblock this.

However, in testing on our 16-core computer, I'm seeing that 'make check-parallel' fails often on:
 `frisbee`,  `posix-mq1`, `posix-mq2`,  `vim`,  `screen`
I'm no sure why those tests are failing, but it also happens with `make AUTOTEST=--slow check-parallel'.

We can push this in now, and test in the main branch (since no one uses 'make check-parallel' normally yet), or else we can keep this purely as a PR, and ask people to do a 'git fetch' to test this PR and try to debug it.  I'm not sure when I will have time to continue debugging.  Any thoughts on the best course of action?

Any thoughts on the best course of action?

gc00 commented:
@rohgarg, I tested on our 16-core computer _without_ `make check-parallel`.  The ordinary test is also failing on things like `posiz-mq1`.  Had you said that you were seeing this test fail in a recent commit (e.g., on DMTCP-3.0)?  If so, do you have a diagnosis of this?

gc00 commented:
I diagnosed the bug.  I place the following three tests first in `autotest.py`:
`posix-mq1` and `posix-mq1` (again).

After that, even the first posix-mq1 fails.  So, there appears to be some sort of race, and there is interference when these process run at that same time.  Since posix-mq1 is using `"/dmtcp-mq-%s"`, there is probably a name clash.  `posix=mq2` seems to also use this name.

So, we have to re-write some of the tests to avoid name clashes with other tests.  In principle (but less likely), we might even see this bug if posix-mq2 is placed after posix-mq1 and any names are not cleaned up by the O/S fast enough between tests.

gc00 committed:
Change test/posix-mq[12] to not use same names

* This is a prerequisite for 'make check-parallel'

rohgarg commented:
@gc00: I have noticed that screen and posix-mq2 are failing on dekaksi, but I haven't had time to investigate the root cause.

> After that, even the first posix-mq1 fails. So, there appears to be some sort of race, and there is interference when these process run at that same time. Since posix-mq1 is using "/dmtcp-mq-%s", there is probably a name clash. posix=mq2 seems to also use this name.
> So, we have to re-write some of the tests to avoid name clashes with other tests.

The weird thing is that it only affects the 3.0 branch. The other two branches -- 2.4 and 2.5 -- are *not* affected by this issue.

gc00 commented:
So, I found one difference between the 2.5 branch and the 3.0/master branch.  The 2.5 branch configure sets `TEST_POSIX_MQ="no"`.  So, we're not doing this problematic test there.

Separately, I renamed the names on `posix-mq1` and that does not fix the bug.  Strangely, sometimes posix-mq1 fails in a sequential test and sometimes it doesn't.  In a parallel test posix-mq1 almost alqays fails.

So, there's still a mystery to be solved here.

gc00 commented:
I just did `posix-mq1` manually, and even without DMTCP, I can see:

> % test/posix-mq1
mq_open() failed: No such file or directory
mq_open() failed: Too many open files

So, the `posix-mq1` test is not cleaning up after itself properly when it crashes.

gc00 commented:
And I think a lot of the problem  is that if the application calls `mq_unlink()` to unlink the file before the checkpoint, then we get:

> [41000] ERROR at fileconnection.cpp:832 in drain; REASON='JASSERT(fd != -1) failed'
     _name = /dmtcp-mq1-gene
     (strerror((*__errno_location ()))) = No such file or directory
posix-mq1 (41000): Terminating.

Even 

Apparently, we try to call `mq_open()` by name to open the mq and drain it.  But that name has already been unlinked by the application, and we had never opened our own private file descriptor, `mqdes`, to use instead of re-opening it.  In our wrapper for `mq_open()`, we should have dup'ed the fd, and then we should close our copy when the user calls `mq_close()` on the original fd.

Even stranger, when I try do do things manually, I get:
> ckpt_posix-mq1_1d4a852a5f139a6-40000-8461cc30f7ac1.dmtcp
ckpt_posix-mq1_1d4a852a5f139a6-41000-58810416.dmtcp

And also, if I let it try to create a second checkpoint (launch/ckpt/resume/ckpt) then it hangs.
So, there are some very strange things happening with this `posix-mq1` test.

gc00 commented:
And I've now removed the tests for `vim` and `screen` when using `make check-parallel`.  (Those tests are special cases, and ordinary sequential `make check` will still call them.  Also, I modified the `posiz-mq1` and `posix-mq2` tests and they work now.  This PR should now be ready to go, once @karya0 unblocks it.  :-)

karya0 commented on test/posix-mq1.c:
```diff
@@ -62,13 +62,19 @@ parent(const char *mqname)
     msg_snd(mqdes, i);
     sleep(1);
+    if (!unlinked) {
+      mq_unlink(mqname); /* parent and child will continue to use mqname */
```
Putting `mq_unlink` changes the semantics of this test. In the previous version, we would ckpt/restore "unlinked" message-queues, whereas in this one, the first checkpoint will be without unlinking. I guess, we can add another test later.
